### PR TITLE
Add error handling in MasterWorker mode

### PR DESCRIPTION
**Pull request by VirrageS**

### CODE REVIEW

File: torch/lib/THD/master_worker/master/Master.cpp [hunk 1]
```diff
@@ -4,29 +4,46 @@
...
 
 using namespace thd;
+using namespace thd::master;
 
 bool THDMasterWorkerInit(THDChannelType channel_type) {
   if (!THDProcessGroupInit(channel_type)) return false;
 
   if (dataChannel->getRank() > 0) {
-    THDWorkerMain();
+    // Worker initialization. Can fail at start but then goes into inifity loop.
```
Comment (0mp) [inline torch/lib/THD/master_worker/master/Master.cpp]: Typo: infinite loop.

File: torch/lib/THD/master_worker/master/Master.cpp [hunk 2]
```diff
@@ -4,29 +4,46 @@
...
+    return THDWorkerMain();
+  }
+
+  // Master initilization
+  masterCommandChannel.reset(new MasterCommandChannel());
+  bool ok = masterCommandChannel->init();
```
Comment (0mp) [inline torch/lib/THD/master_worker/master/Master.cpp]: Is there a reason for this `ok` variable? (You don't use this structure in other files, for example:  torch/lib/THD/master_worker/worker/Worker.cpp)

**Review by 0mp — APPROVED** [review]
+1

File: torch/lib/THD/master_worker/master/Master.cpp [hunk 3]
```diff
@@ -4,29 +4,46 @@
...
 bool THDMasterWorkerInit(THDChannelType channel_type) {
   if (!THDProcessGroupInit(channel_type)) return false;
 
   if (dataChannel->getRank() > 0) {
-    THDWorkerMain();
+    // Worker initialization. Can fail at start but then goes into inifity loop.
+    return THDWorkerMain();
+  }
+
+  // Master initilization
```
Comment (VirrageS) [inline torch/lib/THD/master_worker/master/Master.cpp]: s/initilization/initialization

**Comment by 0mp** [comment]
I've got a compilation error on Ubuntu 16.10:

```
Scanning dependencies of target THD
[  2%] Building CXX object CMakeFiles/THD.dir/base/ChannelUtils.cpp.o
[  5%] Building CXX object CMakeFiles/THD.dir/base/DataChannel.cpp.o
[  8%] Building CXX object CMakeFiles/THD.dir/base/DataChannelRequest.cpp.o
[ 11%] Building CXX object CMakeFiles/THD.dir/base/TensorDescriptor.cpp.o
[ 14%] Building CXX object CMakeFiles/THD.dir/base/data_channels/DataChannelMPI.cpp.o
[ 17%] Building CXX object CMakeFiles/THD.dir/base/data_channels/DataChannelTCP.cpp.o
[ 20%] Building CXX object CMakeFiles/THD.dir/process_group/Collectives.cpp.o
[ 22%] Building CXX object CMakeFiles/THD.dir/process_group/General.cpp.o
[ 25%] Building CXX object CMakeFiles/THD.dir/master_worker/common/ByteArray.cpp.o
[ 28%] Building CXX object CMakeFiles/THD.dir/master_worker/common/CommandChannel.cpp.o
[ 31%] Building CXX object CMakeFiles/THD.dir/master_worker/common/RPC.cpp.o
[ 34%] Building CXX object CMakeFiles/THD.dir/master_worker/master/Master.cpp.o
[ 37%] Building CXX object CMakeFiles/THD.dir/master_worker/master/State.cpp.o
[ 40%] Building CXX object CMakeFiles/THD.dir/master_worker/master/THDRandom.cpp.o
[ 42%] Building CXX object CMakeFiles/THD.dir/master_worker/master/THDStorage.cpp.o
[ 45%] Building CXX object CMakeFiles/THD.dir/master_worker/master/THDTensor.cpp.o
[ 48%] Building CXX object CMakeFiles/THD.dir/master_worker/worker/Dispatch.cpp.o
[ 51%] Building CXX object CMakeFiles/THD.dir/master_worker/worker/Worker.cpp.o
[ 54%] Linking CXX shared library libTHD.so
[ 54%] Built target THD
Scanning dependencies of target test_data_channel_collectives
[ 57%] Building CXX object CMakeFiles/test_data_channel_collectives.dir/test/data_channel_collectives.cpp.o
[ 60%] Linking CXX executable test_data_channel_collectives
libTHD.so: undefined reference to `thd::worker::execute[abi:cxx11](std::unique_ptr<thd::rpc::RPCMessage, std::default_delete<thd::rpc::RPCMessage> >)'
collect2: error: ld returned 1 exit status
CMakeFiles/test_data_channel_collectives.dir/build.make:97: recipe for target 'test_data_channel_collectives' failed
make[2]: *** [test_data_channel_collectives] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/test_data_channel_collectives.dir/all' failed
make[1]: *** [CMakeFiles/test_data_channel_collectives.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
```

File: torch/lib/THD/master_worker/master/Master.cpp (hunk 2 above)
Comment (VirrageS) [inline torch/lib/THD/master_worker/master/Master.cpp]: So in `init` function we make connection with workers. Only if this stage was successful we can start `masterErrorThread`. So before starting this thread we need to make sure everything is okay.

In worker there is no such thing like `errorThread` so there is no need for checking that.

Comment (VirrageS) [reply torch/lib/THD/master_worker/master/Master.cpp]: No, there is no need for that.

**Comment by 0mp** [comment]
The error does not happen in my fresh Ubuntu 16.10 installation in VirtualBox.

File: torch/lib/THD/master_worker/common/CommandChannel.cpp [hunk 4]
```diff
@@ -43,90 +44,152 @@ std::unique_ptr<rpc::RPCMessage> receiveMessage(int socket) {
...
 MasterCommandChannel::MasterCommandChannel()
   : _rank(0)
+  , _poll_events(nullptr)
 {
   rank_type world_size;
   std::tie(_port, world_size) = load_master_env();
 
-  _sockets.resize(world_size);
-  std::fill(_sockets.begin(), _sockets.end(), -1);
+  _connections.resize(world_size);
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/CommandChannel.cpp]: You can initialize that in the member initialization list.

File: torch/lib/THD/master_worker/common/CommandChannel.cpp [hunk 5]
```diff
@@ -84,19 +87,58 @@ bool MasterCommandChannel::init() {
 }
 
 void MasterCommandChannel::sendMessage(std::unique_ptr<rpc::RPCMessage> msg, int rank) {
+  // Throw error which we get from worker.
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/CommandChannel.cpp]: Throw error received from a worker

File: torch/lib/THD/master_worker/common/CommandChannel.cpp [hunk 6]
```diff
@@ -84,19 +87,58 @@ bool MasterCommandChannel::init() {
...
+  if (thd::master::THDState::s_error != "") {
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/CommandChannel.cpp]: I'd make it a `unique_ptr`. Null checks looks cleaner than empty string comparison

File: torch/lib/THD/master_worker/common/CommandChannel.cpp [hunk 7]
```diff
@@ -84,19 +87,58 @@ bool MasterCommandChannel::init() {
...
+    _poll_events.reset(new struct pollfd[_sockets.size()]);
+    for (std::size_t rank = 0; rank < _sockets.size(); ++rank) {
+      _poll_events[rank] = {
+        .fd = _sockets[rank],
+        .events = POLLIN
+      };
+    }
+  }
+
+  // cleanup
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/CommandChannel.cpp]: that comment is unnecessary

File: torch/lib/THD/master_worker/common/CommandChannel.cpp [hunk 8]
```diff
@@ -84,19 +87,58 @@ bool MasterCommandChannel::init() {
...
+  for (std::size_t rank = 0; rank < _sockets.size(); ++rank) {
+    _poll_events[rank].revents = 0;
+  }
+
+  SYSCHECK(::poll(_poll_events.get(), _sockets.size(), -1)) // infinite timeout
+  for (std::size_t rank = 0; rank < _sockets.size(); ++rank) {
+    if (this->_poll_events[rank].revents == 0)
+      continue;
+
+    /* If `POLLIN` is not set or there is different flag set it means that
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/CommandChannel.cpp]: I think we don't have to explain how poll works here

File: torch/lib/THD/master_worker/common/CommandChannel.cpp [hunk 9]
```diff
@@ -84,19 +87,58 @@ bool MasterCommandChannel::init() {
...
+    std::uint64_t error_length;
+    recv_bytes<std::uint64_t>(_poll_events[rank].fd, &error_length, 1);
+
+    std::unique_ptr<char[]> error(new char[error_length + 1]);
+    recv_bytes<char>(_poll_events[rank].fd, error.get(), error_length);
+    return std::make_tuple(rank, std::string(error.get(), error_length));
   }
 
-  return receiveMessage(_sockets[rank]);
+  throw std::runtime_error("Failed to receive error from worker");
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/CommandChannel.cpp]: Wait, why do you always throw here? It seems that you can't exit this function without throwing an error, or returning one received from the worker.

File: torch/lib/THD/master_worker/master/Master.cpp [hunk 10]
```diff
@@ -4,29 +4,50 @@
...
 
+#include <string>
+#include <thread>
 
 namespace thd {
-
 namespace master {
 
 std::unique_ptr<MasterCommandChannel> masterCommandChannel;
+std::thread masterErrorThread;
```
Comment (apaszke) [inline torch/lib/THD/master_worker/master/Master.cpp]: I think it's cleaner to make this thread a member of MasterCommandChannel. It will start it once it's initialized.

File: torch/lib/THD/master_worker/worker/Worker.h [hunk 11]
```diff
@@ -2,4 +2,4 @@
 
 #include "../../THD.h"
 
-THD_API void THDWorkerMain();
+THD_API bool THDWorkerMain();
```
Comment (apaszke) [inline torch/lib/THD/master_worker/worker/Worker.h]: Why change that? It seems that the only value ever returned in `false`.

File: torch/lib/THD/test/command_channel_smoke.cpp [hunk 12]
```diff
@@ -24,18 +25,8 @@ void init_worker(const int& rank, const std::string& master_addr) {
...
-      sizeof("hello to master from worker ") - 1);
-  arr.append(std::to_string(rank).c_str(), std::to_string(rank).size());
-
-  fprintf(stderr, "worker %d: about to send a message\n", rank);
-
-  auto rpc_msg = std::unique_ptr<rpc::RPCMessage>(new rpc::RPCMessage(std::move(arr)));
-  channel->sendMessage(std::move(rpc_msg));
-
-  fprintf(stderr, "worker %d: sent message\n", rank);
+  // Send error
```
Comment (apaszke) [inline torch/lib/THD/test/command_channel_smoke.cpp]: you don't say 😄

**Review by apaszke — CHANGES_REQUESTED** [review]

**Review by 0mp — APPROVED** [review]
+ I fixed the problem -> I got a fresh clone and the problem with `libTHD.so` disappeared.

File: torch/lib/THD/master_worker/common/CommandChannel.cpp [hunk 13]
```diff
@@ -80,23 +84,88 @@ bool MasterCommandChannel::init() {
    // close listen socket
   ::close(_sockets[0]);
   _sockets[0] = -1;
+
+  std::thread error_thread(&MasterCommandChannel::errorHandler, this);
+  error_thread.detach();
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/CommandChannel.cpp]: I think this solution is unsafe. You're detaching the thread, but it keeps reading variables that have their lifetimes bound to the CommandChannel. I think it's possible that the destructor will be called and the object will be freed, but the thread will not manage to notice the state transition. Don't detach it and figure out a way to give it a different stop signal.

File: torch/lib/THD/master_worker/common/CommandChannel.cpp [hunk 14]
```diff
@@ -43,15 +43,23 @@ std::unique_ptr<rpc::RPCMessage> receiveMessage(int socket) {
...
 
-  _sockets.resize(world_size);
-  std::fill(_sockets.begin(), _sockets.end(), -1);
+  _sockets.assign(world_size, -1);
 }
 
 MasterCommandChannel::~MasterCommandChannel() {
+  _exiting = true;
+
+  if (_started)
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/CommandChannel.cpp]: you can replace that flag with `_error_thread.joinable()`

File: torch/lib/THD/master_worker/common/CommandChannel.cpp [hunk 15]
```diff
@@ -80,23 +88,91 @@ bool MasterCommandChannel::init() {
...
+      _poll_events[rank].fd = -1; // mark worker as ignored
+      return std::make_tuple(rank, "connection with worker has been closed");
+    }
+
+    try {
+      // receive error
+      std::uint64_t error_length;
+      recv_bytes<std::uint64_t>(_poll_events[rank].fd, &error_length, 1);
+
+      std::unique_ptr<char[]> error(new char[error_length + 1]);
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/CommandChannel.cpp]: is the array guaranteed to be zero-filled?

File: torch/lib/THD/master_worker/common/CommandChannel.hpp [hunk 16]
```diff
@@ -14,12 +18,20 @@ struct MasterCommandChannel {
...
+  std::tuple<rank_type, std::string> recvError();
+  void errorHandler();
+
   rank_type _rank;
   std::vector<int> _sockets;
+  std::unique_ptr<struct pollfd[]> _poll_events;
+
+  bool _started; // indicates whether error handler thread has started
+  std::atomic_bool _exiting; // informs error handler thread that we are exiting
+  std::unique_ptr<std::string> _error;
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/CommandChannel.hpp]: Actually how about this solution: Open an unnamed pipe and use it in a 0th descriptor of poll. When the channel is destructed write to it, and this will wake up the error thread. Once it sees that it got some data there, it will know that it's an exit message.

File: torch/lib/THD/master_worker/master/Master.cpp [hunk 17]
```diff
@@ -4,29 +4,36 @@
...
 } // namespace thd
 
 using namespace thd;
+using namespace thd::master;
 
 bool THDMasterWorkerInit(THDChannelType channel_type) {
   if (!THDProcessGroupInit(channel_type)) return false;
 
   if (dataChannel->getRank() > 0) {
+    /* Worker initialization. Can fail at start but then goes into infinite loop
```
Comment (apaszke) [inline torch/lib/THD/master_worker/master/Master.cpp]: can fail at start?

File: torch/lib/THD/master_worker/master/Master.cpp [hunk 18]
```diff
@@ -4,29 +4,36 @@
...
+     * in which waits for commands from master.
+     */
     THDWorkerMain();
+    return false;
   }
 
-  // TODO: initialize master
-  thd::master::masterCommandChannel.reset(new MasterCommandChannel());
+  /* Master initialization. We need to make sure that all connections which
```
Comment (apaszke) [inline torch/lib/THD/master_worker/master/Master.cpp]: How is this comment relevant to what's in the code? It looks like a plain constructor with no synchronization with any threads. Maybe consider moving it elsewhere.

File: torch/lib/THD/master_worker/master/Master.hpp [hunk 19]
```diff
@@ -9,7 +9,6 @@ namespace thd {
 namespace master {
 
 extern std::unique_ptr<MasterCommandChannel> masterCommandChannel;
-extern uint64_t nextTensorId;   //TODO wrap it somehow to auto-increment
```
Comment (apaszke) [inline torch/lib/THD/master_worker/master/Master.hpp:12]: is this part of the state or what?

File: torch/lib/THD/test/command_channel_smoke.cpp [hunk 20]
```diff
@@ -24,26 +24,18 @@ void init_worker(const int& rank, const std::string& master_addr) {
...
   auto msg = channel->recvMessage();
   std::string expected = std::string("hello to worker ") +
       std::to_string(rank) + " from master";
   fprintf(stderr, "Worker %d: received '%.*s'\n", rank,
       (int)msg.get()->bytes().length(), msg.get()->bytes().data());
   assert(expected.compare(msg.get()->bytes().to_string()) == 0);
+
+  /*
+   * We need to sleep for a while because instant exiting will cause closing
+   * the socket and in result error in `errorHandler` function.
```
Comment (apaszke) [inline torch/lib/THD/test/command_channel_smoke.cpp]: whaaaaaat?

File: torch/lib/THD/master_worker/master/Master.hpp (hunk 19 above)
Comment (VirrageS) [reply torch/lib/THD/master_worker/master/Master.hpp:12]: It is not used anywhere in project.

File: torch/lib/THD/master_worker/master/Master.cpp (hunk 18 above)
Comment (VirrageS) [reply torch/lib/THD/master_worker/master/Master.cpp]: Oh, I missed that when I moved error_thread to `CommandChannel`. Thanks for catching that!

File: torch/lib/THD/master_worker/master/Master.cpp (hunk 17 above)
Comment (VirrageS) [reply torch/lib/THD/master_worker/master/Master.cpp]: Yes, if `init` function inside `THDWorkerMain` will not succeed it will 'fail'.

Comment (apaszke) [reply torch/lib/THD/master_worker/master/Master.cpp]: idk, I read that sentence as "it can fail, but then *always* goes into an infinite loop".

Comment (apaszke) [reply torch/lib/THD/master_worker/master/Master.cpp]: It's obvious that it can fail, but it's the end of the program then.

File: torch/lib/THD/master_worker/common/CommandChannel.cpp (hunk 15 above)
Comment (VirrageS) [reply torch/lib/THD/master_worker/common/CommandChannel.cpp]: Should we care about that?

File: torch/lib/THD/master_worker/master/Master.cpp (hunk 17 above)
Comment (VirrageS) [reply torch/lib/THD/master_worker/master/Master.cpp]: Oh, right.. Sorry. I will fix that.

File: torch/lib/THD/master_worker/common/CommandChannel.cpp (hunk 15 above)
Comment (apaszke) [reply torch/lib/THD/master_worker/common/CommandChannel.cpp]: well, yes. You never null terminate the memory manually.

File: torch/lib/THD/test/command_channel_smoke.cpp (hunk 20 above)
Comment (VirrageS) [reply torch/lib/THD/test/command_channel_smoke.cpp]: Hmm?

Comment (apaszke) [reply torch/lib/THD/test/command_channel_smoke.cpp]: I can't see how's that a problem. Let's discuss that tomorrow.

Comment (VirrageS) [reply torch/lib/THD/test/command_channel_smoke.cpp]: Oh, maybe its not well described. But yes, we have to wait until master will do all the receiving/sending and only then we can exit. It is because when worker exit it will close socket and trigger poll in error thread what then causes exception.

I think, I have an idea to change `sleep` to `conditional_variable` and `mutex` to be sure that we can safely exit instead of relying on sleeping for 'magic' time.

File: torch/lib/THD/master_worker/common/CommandChannel.cpp (hunk 15 above)
Comment (VirrageS) [reply torch/lib/THD/master_worker/common/CommandChannel.cpp]: Ahh, sorry that `+ 1` is totally unnecessary and misleading. Fortunately, string (2 lines below) is created by taking predefined `error_length` so it is not relying on null termination.

File: torch/lib/THD/master_worker/common/CommandChannel.cpp [hunk 21]
```diff
@@ -43,15 +43,24 @@ std::unique_ptr<rpc::RPCMessage> receiveMessage(int socket) {
...
   std::tie(_port, world_size) = load_master_env();
 
-  _sockets.resize(world_size);
-  std::fill(_sockets.begin(), _sockets.end(), -1);
+  _sockets.assign(world_size, -1);
 }
 
 MasterCommandChannel::~MasterCommandChannel() {
+  if (_error_thread.joinable()) {
+    ::write(_error_pipe, "exit", 4);
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/CommandChannel.cpp]: check if the write fails and print a message to stderr mentioning that

File: torch/lib/THD/master_worker/master/Master.cpp [hunk 22]
```diff
@@ -4,29 +4,33 @@
...
 
 using namespace thd;
+using namespace thd::master;
 
 bool THDMasterWorkerInit(THDChannelType channel_type) {
   if (!THDProcessGroupInit(channel_type)) return false;
 
   if (dataChannel->getRank() > 0) {
+    /*
+     * Worker initialization. If initialization succeeds it goes into
```
Comment (apaszke) [inline torch/lib/THD/master_worker/master/Master.cpp]: Either say that it throws an exception otherwise, or just that it goes into an infinite loop.

Comment (VirrageS) [reply torch/lib/THD/master_worker/master/Master.cpp]: I don't quite understand and not sure how I can clarify it more. Worker does not throw exception on failure it just [silently returns](https://github.com/apaszke/pytorch-dist/blob/thd-mw-error-reporting/torch/lib/THD/master_worker/worker/Worker.cpp#L28) (what implies that it doesn't go into infinite loop). So if it returns it just [immediately hits `return false;`](https://github.com/apaszke/pytorch-dist/blob/thd-mw-error-reporting/torch/lib/THD/master_worker/master/Master.cpp#L27) what should be handled properly in some other component which triggered `THDMasterWorkerInit`.

Comment (apaszke) [reply torch/lib/THD/master_worker/master/Master.cpp]: Ok, that makes sense, I thought it throws. Just say it returns `false` otherwise and it should be ok, sorry for that

**Review by apaszke — APPROVED** [review]
